lite/src/verifier.rs

```diff
@@ -0,0 +1,181 @@
+where
+    H: Header,
+    V: ValidatorSet,
+{
+    // ensure the next validators in the header matches what was supplied.
+    if header.next_validators_hash() != next_vals.hash() {
+        return Err(Error::InvalidNextValidatorSet);
+    }
+
+    Ok(())
+}
+
+// Validate the validators and commit against the header.
+fn validate_vals_and_commit<H, V, C>(header: H, commit: &C, vals: &V) -> Result<(), Error>
+where
+    H: Header,
+    V: ValidatorSet,
+    C: Commit,
+{
+    // ensure the validators in the header matches what we expect from our state.
+    if header.validators_hash() != vals.hash() {
+        return Err(Error::InvalidValidatorSet);
+    }
+
+    // ensure the commit matches the header.
+    if header.hash() != commit.header_hash() {
+        return Err(Error::InvalidCommitValue);
+    }
+
+    Ok(())
+}
+
+/// Verify the commit is valid from the given validators for the header.
+pub fn verify<H, V, C>(header: H, commit: C, validators: V) -> Result<(), Error>
+where
+    H: Header,
+    V: ValidatorSet,
+    C: Commit,
+{
+    if let Err(e) = validate_vals_and_commit(header, &commit, &validators) {
+        return Err(e);
+    }
+
+    // ensure that +2/3 validators signed correctly
+    verify_commit_full(&validators, commit)
+}
+
+/// Verify the commit is trusted according to the last validators and is valid
+/// from the current validators for the header.
+pub fn verify_trusting<H, V, C>(
+    header: H,
+    commit: C,
+    last_validators: V,
+    validators: V,
+) -> Result<(), Error>
+where
+    H: Header,
+    V: ValidatorSetLookup,
+    C: Commit,
+{
+    if let Err(e) = validate_vals_and_commit(header, &commit, &validators) {
+        return Err(e);
+    }
+
+    // ensure that +1/3 of last trusted validators signed correctly
+    if let Err(e) = verify_commit_trusting(&validators, &commit) {
+        return Err(e);
+    }
+
+    // ensure that +2/3 of current validators signed correctly
+    verify_commit_full(&validators, commit)
+}
+
+/// Verify that +2/3 of the correct validator set signed this commit.
+/// NOTE: these validators are expected to be the correct validators for the commit.
+fn verify_commit_full<V, C>(vals: &V, commit: C) -> Result<(), Error>
+where
+    V: ValidatorSet,
+    C: Commit,
+{
+    let total_power = vals.total_power();
+    let mut signed_power: u64 = 0;
+
+    let vals_vec = vals.into_vec();
+    let commit_vec = commit.into_vec();
+
+    if vals_vec.len() != commit_vec.len() {
+        return Err(Error::InvalidCommitLength);
+    }
+
+    // The vals and commit have a 1-to-1 correspondance.
+    // This means we don't need the validator IDs or to do any lookup,
+    // we can just zip the iterators.
+    let vals_iter = vals_vec.into_iter();
+    let commit_iter = commit_vec.into_iter();
+    for (val, vote_opt) in vals_iter.zip(commit_iter) {
+        // skip absent and nil votes
+        // NOTE: do we want to check the validity of votes
```

Comment:
if we're going to punish for them, then yes?

Reply:
We can't necessarily slash for them, because if they're not valid we don't know who signed them. We could "punish" peers if they send us commits with bad nil votes by just disconnecting from them, but it's not clear this actually poses any real risk to us and is worth doing since it seems we can just ignore them.